File: WgConnect/src/com/wgtools/Set.java
/*
 * Copyright 2024 wgconnect@proton.me. All Rights Reserved.
 *
 * WgConnect is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * WgConnect is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with WgConnect.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.wgtools;

import com.wgconnect.core.util.WgConnectLogger;

import java.io.IOException;

import org.apache.commons.lang3.ArrayUtils;

import picocli.CommandLine;
import picocli.CommandLine.Parameters;

/**
 * Set
 * 
 * @author: wgconnect@proton.me
 */
@CommandLine.Command(name = Set.COMMAND, description = Set.DESCRIPTION)
public class Set extends WgSubcommand {
    
    private static final WgConnectLogger log = WgConnectLogger.getLogger(Set.class);    

    public static final String COMMAND = "set";
    public static final String DESCRIPTION = "Change the current configuration, add peers, remove peers, or change peers";
    
    @CommandLine.Option(names = {"-h", "--help"}, defaultValue = "false")
    private boolean helpRequested;
    
    @Parameters(arity="2..*")
    String[] params;
    
    public void showUsage() {
        try {
            parent.getCommandOutputStream().write(
                String.format("Usage: %s %s <interface> [listen-port <port>] [fwmark <mark>] [private-key <file path>]" +
                    " [peer <base64 public key> [remove] [preshared-key <file path>] [endpoint <ip>:<port>] " + 
                    "[persistent-keepalive <interval seconds>] [allowed-ips <ip1>/<cidr1>[,<ip2>/<cidr2>]...] ]...\n",
            parent.spec.commandLine().getCommandName(), Set.COMMAND).getBytes());
        } catch (IOException ex) {
            log.error("Set showUsage exception: " + ex);
        }
    }
    
    @Override
    public void run() {
        parent.setCommandExitCode(0);
        if (helpRequested) {
            showUsage();
            return;
        }

        String[] args = { Set.COMMAND };
        if (params != null)
            args = ArrayUtils.addAll(args, params);
        Object[] results = wgCommand(args);
        parent.setCommandExitCode((Integer)results[0]);
        parent.setCommandOutputString((String)results[1]);
        parent.setCommandErrorString((String)results[2]);
    }
}

File: WgConnect/wireguard-tools/src/set.c
// SPDX-License-Identifier: GPL-2.0
/*
 * Copyright (C) 2015-2020 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include "containers.h"
#include "config.h"
#include "ipc.h"
#include "subcommands.h"
#include "jni-wg.h"

int set_main(int argc, const char *argv[])
{
	struct wgdevice *device = NULL;
	int ret = 1;

	if (argc < 3) {
		jni_err_printf("Usage: %s %s <interface> [listen-port <port>] [fwmark <mark>] [private-key <base64 private-key>] [peer <base64 public key> [remove] [preshared-key <base64 key>] [endpoint <ip>:<port>] [persistent-keepalive <interval seconds>] [allowed-ips <ip1>/<cidr1>[,<ip2>/<cidr2>]...] ]...\n", PROG_NAME, argv[0]);
		return 1;
	}

	device = config_read_cmd(argv + 2, argc - 2);
	if (!device)
		goto cleanup;
	strncpy(device->name, argv[1], IFNAMSIZ -  1);
	device->name[IFNAMSIZ - 1] = '\0';

	if (ipc_set_device(device) != 0) {
		jni_perror("Unable to modify interface", errno);
		goto cleanup;
	}

	ret = 0;

cleanup:
	free_wgdevice(device);
	return ret;
}
